src/main/scala/com/yotpo/metorikku/instrumentation/KafkaLagWriter.scala

```diff
@@ -0,0 +1,41 @@
+package com.yotpo.metorikku.instrumentation
+
+import java.util
+
+import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import org.apache.kafka.clients.consumer.{KafkaConsumer, OffsetAndMetadata}
+import org.apache.kafka.common.TopicPartition
+import org.apache.spark.sql.streaming.StreamingQueryListener
+import org.apache.spark.sql.streaming.StreamingQueryListener._
+
+class KafkaLagWriter(kafkaConsumer: KafkaConsumer[String, String], topic: String) extends StreamingQueryListener {
+  private val consumer = kafkaConsumer
+  @transient lazy val log = org.apache.log4j.LogManager.getLogger(this.getClass)
+
+  def onQueryStarted(event: QueryStartedEvent): Unit = {
+  }
+
+  def onQueryTerminated(event: QueryTerminatedEvent): Unit = {
+
+  }
+
+  def onQueryProgress(event: QueryProgressEvent): Unit = {
+    val om = new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    om.registerModule(DefaultScalaModule)
+    event.progress.sources.foreach(source => {
+      val jsonOffsets = om.readValue(source.endOffset, classOf[Map[String, Map[String, Int]]])
+      jsonOffsets.keys.filter(key => key == topic)
+        .foreach(topic => {
+          val data = jsonOffsets(topic)
```

Comment:
Can you check if empty and write some error?

Reply:
👍